### PR TITLE
Document index structure

### PR DESCRIPTION
**Pull request by pmezard**
Here is a sketch of my understanding of the index structure. As mentioned in the commit message, I usually avoid document internals but core data structures help understanding and communicating a lot.

As usual I second-guessed everything so please correct any mistake. Search algorithms are not mentioned either as I did not check the code exhaustively, only just enough to avoid talking complete nonsense.

While I am there, a couple of questions, some I asked on IRC already:
- Is there any gain in duplicating the field index and position in term frequency vectors? The field index is stored in the row key and the position might be deduced from the element index right? Or maybe there is no gain in removing them either.
- What is the purpose of PrefixCodedInt64? It is reminiscent to stuff I understood lucene was doing to index and query numeric range, by refining requests precision while searching, but that is not completely clear to me. Also, the

```
nChars := (((63 - shift) * 37) >> 8) + 1 
```

is cryptic to me. I would have expected something like:

```
nChars := ((63 - shift + 6)/7
```

since the high order bit is discarded and remaining ones split on 7-bits boundaries.

### CODE REVIEW

File: content/docs/Index-Structure.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+weight = 30
+parent = 'devguide'
+
++++
+
+Bleve default index, called "upside_down" is stored in a single key/value store
+table. The store must be able to enumerate its entries starting at a given key,
+in lexicographic byte order.
+
+Index key and values are handled as byte arrays and are called rows (see `index/upside_down/row.go` for details). To store different row types in a single table, bleve prefixes their keys with a single byte, for instance the term frequency keys start with a 'f'. The following sections describe the data structures stored in the index with pseudo-Go code for value layout.
```
Comment (mschoch) [inline content/docs/Index-Structure.md]: term frequencies rows are prefixed with 't' not 'f'.  field definition rows are prefixed with 'f'.

File: content/docs/Index-Structure.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+Key: (`"d"`, `fieldIndex`, `term`)
+
+Value:
+```
+struct {
+	Count uint64
+}
+```
+
```
Comment (mschoch) [inline content/docs/Index-Structure.md:55]: Just to further clarify, the count here is the number of documents that use the term, not the total number of occurrences of the term.

File: content/docs/Index-Structure.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		Start          uint64
+		End            uint64
+	}
+}
+```
+
+Term frequency rows store statistics about a term in a document field. Term occurrences can be recorded as well.
+
+`Freq` is the number of occurrences of the term in this field.
+`Norm` is 1/sqrt(number of tokens in this field).
```
Comment (mschoch) [inline content/docs/Index-Structure.md]: For further background, the norm is a factor used to help balance scoring of shorter documents which naturally have lower term frequencies.  The value is use at scoring time, but can be computed now at index time.  Future versions may make normalization optional, currently it is not.

File: content/docs/Index-Structure.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+`Vectors` is filled only if `IncludeTermVectors` was set in the field mapping. Each element describes a single occurrence of the term in the source document. `Field` is the field index, and is the same as in the row key. Because documents are structured as trees and intermediate elements can be arrays or slices, the field index is sometimes not enough to resolve a value. The indices of the value or some of its ancestors into those arrays or slices are also required. They are stored in `ArrayPositions`. `Pos` tells the entry is the `Pos`-th occurrence of the term in the field (starting at 1). `Start` and `End` are the byte offsets of the source of the term in the field value.
```
Comment (mschoch) [inline content/docs/Index-Structure.md]: Possibly worth mentioning that term vectors are required if you want to use the following:
- Phrase Search (needs to compare relative Pos of matches)
- Result Highlighting (needs Start and End byte offsets, and ArrayPositions)

**Comment by mschoch** [comment]
The reason we also store field in the term vectors, even though its duplicated from information already in the key is because of composite fields.  So for example, when we search without specifying a field, we might get a match for the field "_all".  So in the term frequency row, the key contains the numeric value corresponding to that field.  But that is not useful if we want to display/highlight the results, we still need to know which field the term usage actually occurred in.  So, in the term vector information, we keep the actual field it occurred in and not "_all".

**Comment by mschoch** [comment]
Regarding the PrefixCodedInt64, yes its used to encode a single number as multiple terms, at different precisions.  The allow a numeric range search to scan fewer rows, though the tuning of this is not very optimal at the moment (I think we way-over-generate terms currently, but I haven't measured it).

Regarding the exact formula in question, it was a port from this Lucene function: https://github.com/apache/lucene-solr/blob/d31c9525b9607efd32cd92d47d1a6e6457bb84d5/lucene/core/src/java/org/apache/lucene/util/NumericUtils.java#L144-L161

File: content/docs/Index-Structure.md (hunk 1 above)
Comment (pmezard) [inline content/docs/Index-Structure.md]: woops, fixed (don't worry if you read "fixed" without a fix, I use that as a reminder, the commit will come eventually).

File: content/docs/Index-Structure.md (hunk 2 above)
Comment (pmezard) [inline content/docs/Index-Structure.md:55]: Fixed.

File: content/docs/Index-Structure.md (hunk 3 above)
Comment (pmezard) [inline content/docs/Index-Structure.md]: Fixed, I added your first sentenced.

File: content/docs/Index-Structure.md (hunk 4 above)
Comment (pmezard) [inline content/docs/Index-Structure.md]: Fixed, I tried to add that to the opening sentence.

**Comment by pmezard** [comment]
I corrected the bit about composite fields (learning what they are at the same time...). Thanks for PrefixCodedInt link.

I do not know how you handle pull-request adjustments. I know I like to review intermediate changes rather than reevaluating everything at each round, at the price of cluttering history. So if you prefer I amend the first changeset I can do that, or if you want me to squash everything when we are done, no problem (you can even do it yourself when merging).

**Comment by pmezard** [comment]
```
// i/7 is the same as (i*37)>>8 for i in 0..63
```

Yeah right, I mostly figured that except for the part that it is exact on 0..63. Still, I wonder how useful is this kind of trick. What I know is it is completely obscure.

**Comment by mschoch** [comment]
I agree it isn't obvious, seems we should either include the comment or switch to the /7 version.

At the moment I'm in favor of switching it to /7 as when I benchmark it I got the following:

```
with benchtime=1s

benchmark                      old ns/op     new ns/op     delta
BenchmarkTestPrefixCoded-4     13750         13755         +0.04%

with benchtime=10s

benchmark                      old ns/op     new ns/op     delta
BenchmarkTestPrefixCoded-4     13965         13692         -1.95%

and with benchtime=1m

benchmark                      old ns/op     new ns/op     delta
BenchmarkTestPrefixCoded-4     13659         13585         -0.54%
```

I'm not sold its really faster, but it doesn't seem like its slower either.

Also, this was with: `nChars := ((63 - shift) / 7) + 1`

but you had proposed: `nChars := ((63 - shift + 6)/7`

Can you explain?

**Comment by pmezard** [comment]
I assume this line computes the number of "7-bits" bytes required to store (63 - shift) bits. If this is the case, current code is incorrect as shown by:

  https://play.golang.org/p/zcAtpPNdJG

See the case "bits: 7", where the current version allocates 2 bytes instead of 1.

But : 
- I can be wrong on what is being computed
- Correctness does not matter anymore, changing the code would break compatibity with existing data.

Your version is equivalent to the current one.

**Comment by mschoch** [comment]
I think the issue is that the original starts with 63, not 64.  If you have a 64 bit value, and shift 63 bits, you still need 1 byte to store the result.  In your case nchars is 0:

```
    bits:  0, current:  1, marty:  1, patrick:  0
```

If I change yours to `patrick := (64 - shift + 6) / 7` it matches for all values.

**Comment by pmezard** [comment]
Yes, I was confused by:

```
sortableBits := int64(uint64(in) ^ 0x8000000000000000)
```

I read that like a `&^` discarding the high-order bit for some unfathomable reason. I suppose flipping it ensures bit lexicographic order is preserved for negative values.

**Pull request merged**